src/main/java/org/sonar/plugins/groovy/codenarc/CodeNarcSourceAnalyzer.java

```diff
@@ -0,0 +1,125 @@
+
+import java.io.File;
+import java.io.FileFilter;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+public class CodeNarcSourceAnalyzer extends AbstractSourceAnalyzer {
+
+  private final Map<File, List<Violation>> violationsByFile = Maps.newHashMap();
+  private final List<File> sourceFiles;
+  private final File baseDir;
+
+  public CodeNarcSourceAnalyzer(List<File> sourceFiles, File baseDir) {
+    this.sourceFiles = sourceFiles;
+    this.baseDir = baseDir;
+  }
+
+  @Override
+  public Results analyze(RuleSet ruleSet) {
+    Multimap<File, FileResults> resultsByFileByDirectory = processFiles(ruleSet);
+    return processDirectories(baseDir, resultsByFileByDirectory);
+  }
+
+  private Multimap<File, FileResults> processFiles(RuleSet ruleSet) {
+    Multimap<File, FileResults> results = LinkedListMultimap.create();
+    for (File file : sourceFiles) {
+      List<Violation> violations = collectViolations(new SourceFile(file), ruleSet);
+      violationsByFile.put(file, violations);
+      FileResults result = new FileResults(file.getAbsolutePath(), violations);
+      results.put(file.getParentFile(), result);
+    }
+    return results;
+  }
+
+  private static DirectoryResults processDirectories(File baseDir, Multimap<File, FileResults> resultsByFileByDirectory) {
+    Map<File, DirectoryResults> results = prepopulateDirectoryResults(baseDir, resultsByFileByDirectory);
+
+    for (File directory : resultsByFileByDirectory.keySet()) {
+      DirectoryResults directoryResult = results.get(directory);
+      Collection<FileResults> filesResults = resultsByFileByDirectory.get(directory);
+      for (FileResults fileResults : filesResults) {
+        directoryResult.addChild(fileResults);
+      }
+    }
+    return results.get(baseDir);
+  }
+
+  private static Map<File, DirectoryResults> prepopulateDirectoryResults(File baseDir, Multimap<File, FileResults> resultsByFileByDirectory) {
+    Map<File, DirectoryResults> results = Maps.newHashMap();
+    results.put(baseDir, new DirectoryResults());
+    // add a result by directory for each level from current directory to baseDir
+    for (File directory : resultsByFileByDirectory.keySet()) {
+      File parent = directory;
+      while (!results.containsKey(parent)) {
+        results.put(parent, new DirectoryResults());
+        parent = parent.getParentFile();
+      }
+    }
+
+    // construct parent-child relationship between results
+    for (Entry<File, DirectoryResults> resultByDirectory : results.entrySet()) {
+      DirectoryResults result = resultByDirectory.getValue();
+      for (File subDirectory : subDirectories(resultByDirectory.getKey())) {
+        if (results.containsKey(subDirectory)) {
+          result.addChild(results.get(subDirectory));
+        }
+      }
+    }
+
+    return results;
+  }
+
+  private static File[] subDirectories(File file) {
+    return file.listFiles(new FileFilter() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a named "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)